src/include/detail/graph/nn-descent.h

```diff
@@ -0,0 +1,470 @@
+6. Return to step 2, unless we have already tried all the candidates in the pool
+
+  */
+
+/*
+    NN-descent algorithm
+    1. Start with a random graph (connect each node to 𝑘 random nodes)
+    2. For each node:
+    3. Measure the distance from the node to the neighbors of its neighbors
+    4. If any are closer, then update the graph accordingly, and keep only the 𝑘
+   closest
+    5. If any updates were made to the graph then go back to step 2, otherwise
+   stop
+ */
+
+template <
+    std::integral I,
+    std::integral J,
+    class Distance = sum_of_squares_distance>
+auto nn_descent_0_step(
+    auto&& g, auto&& db, I i, J j, Distance distance = Distance()) {
+  size_t num_updates{0};
+  for (auto&& [_, k] : out_edges(g, j)) {  // _ is dist(j, k)
+    if (k == i) {
+      continue;
+    }
+    auto ik_score = distance(db[i], db[k]);
+    if (g.add_edge(i, k, ik_score)) {
+      ++num_updates;
+    }
+  }
+  for (auto&& k : in_edges(g, j)) {  // _ is dist(j, k)
+    if (k == i) {
+      continue;
+    }
+    auto ik_score = distance(db[i], db[k]);
+    if (g.add_edge(i, k, ik_score)) {
+      ++num_updates;
+    }
+  }
+
+  return num_updates;
+}
+
+template <class Distance = sum_of_squares_distance>
+auto nn_descent_0_step_all(
+    auto&& g, auto&& db, Distance distance = Distance()) {
+  size_t num_updates{0};
+  size_t nvertices{num_vertices(g)};
+  // g.copy_to_update_edges();
+
+  g.build_in_edges();
+
+  for (size_t i = 0; i < nvertices; ++i) {
+    for (auto&& [_, j] : out_edges(g, i)) {
+      num_updates += nn_descent_0_step(g, db, i, j);
+    }
+
+    for (auto&& j : in_edges(g, i)) {
+      num_updates += nn_descent_0_step(g, db, i, j);
+    }
+  }
+  // g.swap_all_update_edges();
+
+  return num_updates;
+}
+
+// Local join
+template <
+    std::integral I,
+    std::integral J,
+    class Distance = sum_of_squares_distance>
+auto nn_descent_1_step(
+    auto&& g, auto&& db, I i, J j, Distance distance = Distance()) {
+  size_t num_updates{0};
+
+  for (auto&& [_, k] : out_edges(g, i)) {  // _ is dist(j, k)
+    if (k <= j) {
+      continue;
+    }
+    auto jk_score = distance(db[j], db[k]);
+    if (g.add_edge(j, k, jk_score)) {
+      ++num_updates;
+    }
+    if (g.add_edge(k, j, jk_score)) {
+      ++num_updates;
+    }
+  }
+
+  for (auto&& k : in_edges(g, i)) {  // _ is dist(j, k)
+    if (k <= j) {
+      continue;
+    }
+    auto jk_score = distance(db[j], db[k]);
+    if (g.add_edge(j, k, jk_score)) {
+      ++num_updates;
+    }
+    if (g.add_edge(k, j, jk_score)) {
+      ++num_updates;
+    }
+  }
+
+  return num_updates;
+}
+
+template <class Distance = sum_of_squares_distance>
+auto nn_descent_1_step_all(
+    auto&& g, auto&& db, Distance distance = Distance()) {
+  size_t num_updates{0};
+  size_t nvertices{num_vertices(g)};
+  // g.copy_to_update_edges();
+
+  g.build_in_edges();
+
+  for (size_t i = 0; i < nvertices; ++i) {
+    for (auto&& [_, j] : out_edges(g, i)) {
+      // print_types(i, j);
```

Comment:
Remove commented lines here and in other parts of the file. If something need to remain add a comment for why it is needed.